### PR TITLE
added all files back

### PR DESCRIPTION
**Pull request by Malarg**

### CODE REVIEW

File: backend/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+WORKDIR /app
+
+COPY go.mod go.sum ./
+
+RUN go mod download
+
+COPY . .
+
+RUN CGO_ENABLED=0 GOOS=linux go build -a -installsuffix cgo -o main ./cmd/app
```
Comment (nikola-jokic) [inline backend/Dockerfile]: I don't think you need `-installsuffix cgo` when CGO_ENABLED=0

File: backend/Makefile [hunk 2]
```diff
@@ -0,0 +1,16 @@
+localConfigName ?= local
+
+run:
+	CONFIG_NAME=$(localConfigName) docker-compose up --build -d
```
Comment (nikola-jokic) [inline backend/Makefile]: Suggestion: Use docker compose instead of docker-compose

File: backend/cmd/app/main.go [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+	ginSwagger "github.com/swaggo/gin-swagger"
+)
+
+// @title Shodo API
+// @version 1.0
+// @description This is methods declaration for Shodo API
+// @host localhost:8080
+// @BasePath /api/v1
+func main() {
+	r := gin.Default()
```
Comment (nikola-jokic) [inline backend/cmd/app/main.go]: It can be confusing to have gin instantiation and swagger here, and internal/app doing rest of it.
Main func is usually used to setup dependencies and pass them to an app. Then, the app should accept those dependencies and start building from there.

Usually, in main, you would instantiate a database, read the config, do initialization and pass it to the app.
The app assumes you already have everything passed instantiated correctly, and it's responsibility would be to define routes and handlers for your project.
Mixing it up is hard to follow

File: backend/internal/app/app.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+	"github.com/gin-gonic/gin"
+	"github.com/go-redis/redis"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"go.uber.org/zap"
+)
+
+func Run(r *gin.Engine) {
+
+	Logger, err := zap.NewProduction()
```
Comment (nikola-jokic) [inline backend/internal/app/app.go]: Variable names should be lowercase :relaxed:

File: backend/internal/app/app.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+	}
+
+	configName := os.Getenv("CONFIG_NAME")
+	config, err := config.Init(configName)
+	if err != nil {
+		fmt.Println(err)
+		panic(err)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
```
Comment (nikola-jokic) [inline backend/internal/app/app.go]: If you want to timeout db instantiation, use a separate function so you will call cancel as soon as you can
This way, if the connection is successful, you will release timers only when the app finishes

File: backend/internal/app/app.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+	defer cancel()
+	mongoUrl := fmt.Sprintf("mongodb://admin:password@%s:27017", config.MongoHost)
```
Comment (nikola-jokic) [inline backend/internal/app/app.go]: Best practice is to move this to config. You already have a MongoHost in config, why not credentials?
You can even go one step further and encode password, but that might not be necessary

File: backend/internal/app/app.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+		users.GET("/", usersHandler.GetAllUsers)
+	}
+	v1.GET("/ping", transport.PingHandler)
+
+	r.Run()
+}
+
+func disconnectMongoDB(ctx context.Context, client *mongo.Client) {
+	if err := client.Disconnect(ctx); err != nil {
+		panic(err)
```
Comment (nikola-jokic) [inline backend/internal/app/app.go]: This is only called when the app is exiting, so panic won't stop the execution. The better approach is to log this error with the stack trace if you need. Usually, panic is used in exceptional situations where you want to end the execution due to an error. Here, you only need to know that disconnect failed.

File: backend/internal/domain/helpers/collections.go [hunk 8]
```diff
@@ -0,0 +1,10 @@
+package helpers
```
Comment (nikola-jokic) [inline backend/internal/domain/helpers/collections.go]: Best practice is to avoid utils/helpers/...
The package name serves for readability also, and helpers does not add any context. It is better to have a small amount of duplication, than a wrong abstraction.
If you need helpers, you should (in this case) provide a `slices` package, and then the call would be:
```go
slices.Contains(s, element)
```
You can even go a step further. You are currently allowing string | int. Why not `comparable` to make it general purpose?

File: backend/internal/domain/helpers/password_hasher.go [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+	"golang.org/x/crypto/bcrypt"
+)
+
+const (
+	DefaultCost int = 10
+)
+
+func HashPassword(password string) (string, error) {
+	if len(password) > 36 {
+		ErrPasswordTooLong := errors.New("password length should not exceed 72 bytes")
```
Comment (nikola-jokic) [inline backend/internal/domain/helpers/password_hasher.go]: variable names should always be lowercase
Question: how do you know that the len(password) == 36 is exactly 72 bytes?
It is utf8 encoded, it can be 36bytes up to 4*36

File: backend/internal/domain/helpers/password_hasher.go [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+		return "", ErrPasswordTooLong
+	}
+
+	bytes, err := bcrypt.GenerateFromPassword([]byte(password), DefaultCost)
```
Comment (nikola-jokic) [inline backend/internal/domain/helpers/password_hasher.go]: Usually, when you return an error, you should return a zero value of a type. The approach is to make value useless if an error occurred.

Here, I think the best approach would be to say:
```go
if err != nil {
    return "", err
}
return string(bytes), nil
```

File: backend/internal/domain/helpers/tokens_generator.go [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+
+	"github.com/golang-jwt/jwt"
+)
+
+const (
+	AccessTokenLifeTime  = 3600
+	RefreshTokenLifeTime = 86400
+)
+
+var Secret = os.Getenv("JWT_SECRET")
```
Comment (nikola-jokic) [inline backend/internal/domain/helpers/tokens_generator.go]: Mixing config and os env can be hard to track
Ideally, you would have a one place where you source everything, and wire it up in the main or some other function.

File: backend/internal/domain/helpers/tokens_generator.go [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+
+func GenerateTokens(userId string) (models.AuthTokens, error) {
```
Comment (nikola-jokic) [inline backend/internal/domain/helpers/tokens_generator.go]: NIT: Common words like id, url, etc should follow the same casing
So you can use either: id or ID, but not Id
Same goes for URL or url, not Url

This is a nit but usually most style guides are following this convention

File: backend/internal/domain/services/authentication_service.go [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+
+	"golang.org/x/net/context"
+)
+
+type AuthenticationService struct {
+	Repository    repository.Users
+	TokensService *TokensService
+}
+
+func (this *AuthenticationService) LogIn(ctx context.Context, request models.LoginUserRequest) (*models.AuthTokens, error) {
```
Comment (nikola-jokic) [inline backend/internal/domain/services/authentication_service.go]: You should never use receiver as `this` or `self`
It is an abbreviation of the type it receives. So here, I'd use `s` here to signal the service.

File: backend/internal/domain/services/registration_service.go [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+func (this *RegistrationService) checkIfCanRegister(ctx context.Context, request models.RegisterUserRequest) error {
+	userExists, err := this.Repository.CheckUserExists(ctx, request.Email)
+	if err != nil {
+		return err
+	}
+
+	if userExists {
+		return errors.New(fmt.Sprintf("user with email %s already exists", request.Email))
```
Comment (nikola-jokic) [inline backend/internal/domain/services/registration_service.go]: Why not `fmt.Errorf`?

File: backend/internal/domain/services/registration_service.go [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+	}
+
+	if len(request.Password) < MinPasswordLength {
+		message := fmt.Sprintf("password length must be greater than %d characters", MinPasswordLength)
+		return errors.New(message)
+	}
+	return nil
+}
+
+func (this *RegistrationService) putNewUserToDb(ctx context.Context, request models.RegisterUserRequest) (string, error) {
```
Comment (nikola-jokic) [inline backend/internal/domain/services/registration_service.go]: A better name would be createUser.

File: backend/internal/domain/services/task_list_service.go [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+import (
+	"context"
+	"net/http"
+	"shodo/internal/domain/helpers"
+	"shodo/internal/repository"
+	"shodo/models"
+)
+
+const (
+	kDefaultTaskListTitle = "Shoppings"
```
Comment (nikola-jokic) [inline backend/internal/domain/services/task_list_service.go]: never name constants with the k prefix
constant names are based on the meaning they represent.

In go, we don't put `k` in front of constants the same way we don't prefix interfaces with `i`

File: backend/internal/domain/services/task_list_service.go [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+	userId, err := helpers.GetUserIdFromToken(userToken)
+	if err != nil {
+		return nil, err
+	}
+
+	return this.TaskListRepository.GetTaskLists(ctx, userId)
+}
+
+func (this *TaskListService) GetTaskList(ctx context.Context, listId *string, userToken string) (models.TaskList, *models.Error) {
+	isListExists, err := this.TaskListRepository.CheckTaskListExists(ctx, *listId)
```
Comment (nikola-jokic) [inline backend/internal/domain/services/task_list_service.go]: You rarely use is... in front of a boolean.
`exists` would be a better name and it reads well

`if exists` is nicer than `if isListExists`
Also, you usually don't put the datatype name in a variable

You already know the type since this is a strongly typed language

File: backend/internal/domain/services/task_list_service.go [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+
+	taskId, err := this.TaskListRepository.AddTaskToList(ctx, listId, task)
+	if err != nil {
+		return nil, &models.Error{Code: http.StatusInternalServerError, Message: err.Error()}
+	}
+
+	return taskId, nil
+}
+
+func (this *TaskListService) RemoveTaskFromList(ctx context.Context, listId *string, taskId *string, userToken string) *models.Error {
```
Comment (nikola-jokic) [inline backend/internal/domain/services/task_list_service.go]: If you need to signal an error, you need to return a type `error`
It can lead to subtle problems later and you should never base your errors on strong types.

You can always inspect them with `errors.Is` or `errors.As`

File: backend/internal/repository/mongo_dto/task_list.go [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+	Title      string             `bson:"title,omitempty"`
+	Owner      primitive.ObjectID `bson:"owner,omitempty"`
+	SharedWith []primitive.ObjectID
+	Tasks      []Task `bson:"tasks,omitempty"`
+}
+
+func (this *TaskList) New() {
+	this.ID = primitive.ObjectID{}
+	this.Title = ""
+	this.Owner = primitive.ObjectID{}
```
Comment (nikola-jokic) [inline backend/internal/repository/mongo_dto/task_list.go]: New like this is basically a reset. So why not call it like that?
New signals creation of a type, and you are not creating a type here :relaxed:

File: backend/tests/requests.go [hunk 20]
```diff
@@ -0,0 +1,153 @@
...
+func (s *APITestSuite) sendJSONRequest(method, url string, request interface{}, authToken *string) (*http.Response, error) {
+	requestBytes, err := json.Marshal(request)
+	if err != nil {
+		return nil, err
+	}
+
+	return s.sendRequest(method, url, requestBytes, authToken)
+}
+
+func (s *APITestSuite) sendRegisterRequest(request models.RegisterUserRequest) (*http.Response, error) {
```
Comment (nikola-jokic) [inline backend/tests/requests.go]: why not use `httptest.ResponseRecorder`?

File: backend/Dockerfile (hunk 1 above)
Comment (Malarg) [reply backend/Dockerfile]: Thank you, removed

File: backend/Makefile (hunk 2 above)
Comment (Malarg) [reply backend/Makefile]: Read about it, removed :)

File: backend/cmd/app/main.go (hunk 3 above)
Comment (Malarg) [reply backend/cmd/app/main.go]: Fixed, moved all initialization into main and rest into app

File: backend/internal/app/app.go (hunk 4 above)
Comment (Malarg) [reply backend/internal/app/app.go]: Thank you, fixed

File: backend/internal/app/app.go (hunk 5 above)
Comment (Malarg) [reply backend/internal/app/app.go]: Fixed)

File: backend/internal/app/app.go (hunk 6 above)
Comment (Malarg) [reply backend/internal/app/app.go]: Moved

File: backend/internal/app/app.go (hunk 7 above)
Comment (Malarg) [reply backend/internal/app/app.go]: Very helpful, fixed

File: backend/internal/domain/helpers/collections.go (hunk 8 above)
Comment (Malarg) [reply backend/internal/domain/helpers/collections.go]: Removed helpers folder

File: backend/internal/domain/helpers/password_hasher.go (hunk 9 above)
Comment (Malarg) [reply backend/internal/domain/helpers/password_hasher.go]: Read `len` documentation. Looks like it returns bytes numbers, not string length. Fixed

File: backend/internal/domain/helpers/password_hasher.go (hunk 10 above)
Comment (Malarg) [reply backend/internal/domain/helpers/password_hasher.go]: Fixed

File: backend/internal/domain/helpers/tokens_generator.go (hunk 12 above)
Comment (Malarg) [reply backend/internal/domain/helpers/tokens_generator.go]: Thank you. You are right, many companies folliwing this conventions. But in several last places i worked we wrote `Id`, `Url` and so on. It's just a habit

File: backend/internal/domain/services/authentication_service.go (hunk 13 above)
Comment (Malarg) [reply backend/internal/domain/services/authentication_service.go]: Fixed)

File: backend/internal/domain/services/registration_service.go (hunk 14 above)
Comment (Malarg) [reply backend/internal/domain/services/registration_service.go]: Fixed

File: backend/internal/domain/services/registration_service.go (hunk 15 above)
Comment (Malarg) [reply backend/internal/domain/services/registration_service.go]: Renamed, Thank you

File: backend/internal/domain/services/task_list_service.go (hunk 16 above)
Comment (Malarg) [reply backend/internal/domain/services/task_list_service.go]: Renamed

File: backend/internal/domain/services/task_list_service.go (hunk 17 above)
Comment (Malarg) [reply backend/internal/domain/services/task_list_service.go]: I prefer user `is` and `are` in front of boolean variables in all languages i worked with :) 
I'm sure that this habit make naming more clear

File: backend/internal/domain/services/task_list_service.go (hunk 18 above)
Comment (Malarg) [reply backend/internal/domain/services/task_list_service.go]: Ok, but how should I tell handler the error code?

File: backend/internal/repository/mongo_dto/task_list.go (hunk 19 above)
Comment (Malarg) [reply backend/internal/repository/mongo_dto/task_list.go]: Removed New at all. It is not used and I don't like mutable objects

File: backend/tests/requests.go (hunk 20 above)
Comment (Malarg) [reply backend/tests/requests.go]: Unfortunately I didn't understood your thought. Can you explain, please, how exactly should I use `httptest.ResponseRecorder`?

File: backend/internal/domain/helpers/tokens_generator.go (hunk 11 above)
Comment (Malarg) [reply backend/internal/domain/helpers/tokens_generator.go]: Ok, thank you got it. I'll move it to main in next iteration of fixing the project :)

File: backend/internal/domain/services/task_list_service.go (hunk 17 above)
Comment (nikola-jokic) [reply backend/internal/domain/services/task_list_service.go]: I know it is a common convention in other languages, but it does not really work with go at least go style guide.
It is the same for getters. In Go, we don't prefix with Get and Set.
If you are retrieving value, you just name it like: `Name()` instead of `GetName()`. For setters, you prefix it with `SetName(name)`.
There is no really a need for is and are because you already know the type of the variable. The same goes with naming collections. It is not `userList`, it is `users`. But you can write it however you would like :relaxed: I'm just here to provide best practices related to go development style :smile:

File: backend/internal/domain/helpers/password_hasher.go (hunk 9 above)
Comment (nikola-jokic) [reply backend/internal/domain/helpers/password_hasher.go]: You can always leverage `utf8.RuneCountInString` to count number of characters :relaxed:

File: backend/internal/domain/services/task_list_service.go (hunk 18 above)
Comment (nikola-jokic) [reply backend/internal/domain/services/task_list_service.go]: If you use errors.As, you can source error type into your structure easily. Something like
```go
var modelErr *models.Error
if errors.As(err, &modelErr) {
// your error is sourced into modelErr so you can inspect it
}
```
Nevertheless, I don't think the model should contain errors. Model should represent the objects of the application. Error signals the incorrect state. So ideally, the package containing the logic should have errors specific to its logic.

Also, the issue I have with this kind of design is that your service knows about the transport layer, which kind of defeats the whole purpose of splitting it into handler/service/db. Handler is responsible for communicating values. Service is responsible for the logic and the db holds the state. If you know that your app is always going to be using http, what is the purpose of the handler/service split?

I know apps that can be triggered either through web sockets or HTTP requests. Then it would re-use service logic, but the error code returned would mean nothing to the websocket (because it is basically just a TCP stream wrapper)

File: backend/tests/requests.go (hunk 20 above)
Comment (nikola-jokic) [reply backend/tests/requests.go]: Oh I see, your tests directory is basically e2e tests :relaxed: 
Ignore me here

File: backend/internal/repository/mongo_dto/task_list.go (hunk 19 above)
Comment (nikola-jokic) [reply backend/internal/repository/mongo_dto/task_list.go]: That is okay, but keep in mind that your performance is going to be hit on a large scale if you are going to copy values each time you pass them to other functions

File: backend/internal/domain/services/task_list_service.go (hunk 18 above)
Comment (Malarg) [reply backend/internal/domain/services/task_list_service.go]: Oh, I got my mistake. Thank you :)
Please check how I changed working with errors for this method. Now layers are less bound between each other. If you approve this, I'll change error handling for all other methods in this app

File: backend/internal/domain/helpers/password_hasher.go (hunk 9 above)
Comment (Malarg) [reply backend/internal/domain/helpers/password_hasher.go]: Thank you :)

Comment (Malarg) [reply backend/internal/domain/helpers/password_hasher.go]: But for this case it's not necessary, because `GenerateFromPassword` works with bytes
